### PR TITLE
feat(icon): implement icon component

### PR DESCRIPTION
**Pull request by schnerd**
Initial implementation for the Icon component.

### [Live Preview](https://deploy-preview-269--baseui.netlify.com/?knob-Size=48px&knob-Color=0&knob-Button%20Kind=primary&selectedKind=Icon&selectedStory=Icon&full=0&addons=1&stories=1&panelRight=0&addonPanel=REACT_STORYBOOK%2Freadme%2Fpanel)

**Progress**
- [x] Basic implementation
- [x] Stories
- [x] Implement all icons (23 / 23)
- [x] Unit Tests

**Questions**

- Should we use names that describe the iconography itself or its semantic meaning? For example, the X icon can be used for both `Delete` and `Close` use cases, so maybe `X` is more appropriate? In theory we could have aliases too.

### CODE REVIEW

File: src/icon/delete-alt.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+LICENSE file in the root directory of this source tree.
+*/
+// @flow
+import * as React from 'react';
+import Icon from './icon';
+import type {IconPropsT} from './types';
+
+export default function IconDeleteAlt(props: IconPropsT) {
+  return (
+    <Icon viewBox="0 0 16 16" title="Delete" {...props}>
```
[the remark this answers is not in the document]
Comment (schnerd) [reply src/icon/delete-alt.js]: Conclusion – yes everything should have viewbox="0 0 24 24" matching the white rectangle in the docs:

image

side effect of this is that the a prop like `size="12"` doesn't necessarily mean your icon will be exactly 12px tall/wide. But the icons are designed in a way to be cohesive from a design perspective without actually all being the same size.

File: src/icon/icon-template.txt [hunk 2]
```diff
@@ -0,0 +1,19 @@
+/*
+Copyright (c) 2018 Uber Technologies, Inc.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+*/
+// @flow
+// BASEUI-GENERATED-REACT-ICON
```
Comment (gergelyke) [inline src/icon/icon-template.txt]: should we also explicitly state that one should not modify it manually?

File: rfcs/icon-component.md [hunk 3]
```diff
@@ -48,7 +48,7 @@ From left-to-right based on the image above:
 ## `Icon` API
 
 * `size: number|string` - Optional, Defaults to `scale500` (12px)
-  Size of element, will be passed to the svg width/height style. Can also be a baseui `theme.sizing` constant like `sizing500`.
+  Size of element, will be passed to the svg width/height style. Can also be a baseui `theme.sizing` constant like `scale500`.
```
Comment (gergelyke) [inline rfcs/icon-component.md]: i know it is not part of this PR, but should we move the usage example up, so it follows the order we have in the rest of the rfcs?

Comment (gergelyke) [reply rfcs/icon-component.md]: also, should we include an example with buttons too?

File: src/icon/icon-template.txt (hunk 2 above)
Comment (schnerd) [reply src/icon/icon-template.txt]: good call, will update.

File: rfcs/icon-component.md (hunk 3 above)
Comment (schnerd) [reply rfcs/icon-component.md]: Good call, will add.

Comment (kahwee) [reply rfcs/icon-component.md]: Wow I really like how the icons looks in your [live preview](https://deploy-preview-269--baseui.netlify.com/?knob-Size=48px&knob-Color=0&knob-Button%20Kind=primary&selectedKind=Icon&selectedStory=Icons%20in%20Button&full=0&addons=1&stories=1&panelRight=0&addonPanel=REACT_STORYBOOK%2Freadme%2Fpanel).

Maybe it's just me but the start and end enhancers in the button seems to have too much whitespace. What do you think @nguyensomniac ?

**Pull request merged**